### PR TITLE
Bbmap/bbduk

### PR DESCRIPTION
**Pull request by MGordon09**
## PR checklist

Closes #XXX 

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: software/bbmap/bbduk/main.nf [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        container "quay.io/biocontainers/bbmap:38.90--he522d1c_1"
+    }
+
+    input:
+    tuple val(meta), path(reads)
+    path contaminants
+    val use_contaminants
+
+    output:
+    tuple val(meta), path('*.trim.fastq.gz'), emit: reads
```
Comment (drpatelh) [inline software/bbmap/bbduk/main.nf:26]:
```suggestion
    tuple val(meta), path('*.fastq.gz'), emit: reads
```

File: software/bbmap/bbduk/main.nf [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    tuple val(meta), path('*.log')          , emit: log
+    path '*.version.txt'                    , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    def raw      = meta.single_end ? "in=${reads[0]}" : "in1=${reads[0]} in2=${reads[1]}"
+    def trimmed  = meta.single_end ? "out=${prefix}.trim.fastq.gz" : "out1=${prefix}_1.trim.fastq.gz out2=${prefix}_2.trim.fastq.gz"
```
Comment (drpatelh) [inline software/bbmap/bbduk/main.nf:34]:
```suggestion
    def trimmed  = meta.single_end ? "out=${prefix}.fastq.gz" : "out1=${prefix}_1.fastq.gz out2=${prefix}_2.trim.fastq.gz"
```

File: software/bbmap/bbduk/main.nf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    def contaminants_fa = use_contaminants ? "ref=$contaminants" : ''
+
```
Comment (drpatelh) [inline software/bbmap/bbduk/main.nf:36]:
```suggestion
```

File: software/bbmap/bbduk/main.nf [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/bbmap:38.90--he522d1c_1"
+    } else {
+        container "quay.io/biocontainers/bbmap:38.90--he522d1c_1"
+    }
+
+    input:
+    tuple val(meta), path(reads)
+    path contaminants
+    val use_contaminants
```
Comment (drpatelh) [inline software/bbmap/bbduk/main.nf:23]:
```suggestion
```

File: software/bbmap/bbduk/main.nf [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    tuple val(meta), path('*.trim.fastq.gz'), emit: reads
+    tuple val(meta), path('*.log')          , emit: log
+    path '*.version.txt'                    , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    def raw      = meta.single_end ? "in=${reads[0]}" : "in1=${reads[0]} in2=${reads[1]}"
+    def trimmed  = meta.single_end ? "out=${prefix}.trim.fastq.gz" : "out1=${prefix}_1.trim.fastq.gz out2=${prefix}_2.trim.fastq.gz"
+    def contaminants_fa = use_contaminants ? "ref=$contaminants" : ''
```
Comment (drpatelh) [inline software/bbmap/bbduk/main.nf:35]:
```suggestion
    def contaminants_fa = contaminants ? "ref=$contaminants" : ''
```

File: tests/config/pytest_software.yml [hunk 6]
```diff
@@ -18,6 +18,10 @@ bandage/image:
   - software/bandage/image/**
   - tests/software/bandage/image/**
 
+bbmap_bbduk:
```
Comment (drpatelh) [inline tests/config/pytest_software.yml:21]:
```suggestion
bbmap/bbduk:
```

File: tests/software/bbmap/bbduk/main.nf [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+include { BBMAP_BBDUK } from '../../../../software/bbmap/bbduk/main.nf' addParams( options: [ 'args' : 'trimq=10 qtrim=r' ] )
+
+workflow test_bbmap_bbduk_single_end {
+
+    input = [ [ id:'test', single_end:true ], // meta map
+              [  file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true) ]
+            ]
+    contaminants = file('contaminants_dummy')
+    use_contaminants = false
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:13]:
```suggestion
```

File: tests/software/bbmap/bbduk/main.nf [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    BBMAP_BBDUK ( input, contaminants, use_contaminants )
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:15]:
```suggestion
    BBMAP_BBDUK ( input, contaminants )
```

File: tests/software/bbmap/bbduk/main.nf [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+workflow test_bbmap_bbduk_paired_end {
+
+    input = [ [ id:'test', single_end:false ], // meta map
+              [  file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
+                 file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
+            ]
+    contaminants = file('contaminants_dummy')
+    use_contaminants = false
+
+    BBMAP_BBDUK ( input, contaminants, use_contaminants )
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:27]:
```suggestion
    BBMAP_BBDUK ( input, [] )
```

File: tests/software/bbmap/bbduk/main.nf [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+workflow test_bbmap_bbduk_se_ref {
+
+    input = [ [ id:'test', single_end:true ], // meta map
+              [  file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true) ]
+
+            ]
+    contaminants = [file(params.test_data['sarscov2']['genome']['transcriptome_fasta'], checkIfExists: true) ] // transciptome file - remove contaminants (*trim.fastq files empty)
+    use_contaminants = true
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:37]:
```suggestion
```

File: tests/software/bbmap/bbduk/main.nf [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+    BBMAP_BBDUK ( input, contaminants, use_contaminants )
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:39]:
```suggestion
    BBMAP_BBDUK ( input, contaminants )
```

File: tests/software/bbmap/bbduk/main.nf [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+workflow test_bbmap_bbduk_paired_end {
+
+    input = [ [ id:'test', single_end:false ], // meta map
+              [  file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
+                 file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
+            ]
+    contaminants = file('contaminants_dummy')
+    use_contaminants = false
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:25]:
```suggestion
```

File: tests/software/bbmap/bbduk/main.nf [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+              [  file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true) ]
+
+            ]
+    contaminants = [file(params.test_data['sarscov2']['genome']['transcriptome_fasta'], checkIfExists: true) ] // transciptome file - remove contaminants (*trim.fastq files empty)
+    use_contaminants = true
+
+    BBMAP_BBDUK ( input, contaminants, use_contaminants )
+}
+
+
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:42]:
```suggestion
```

File: tests/software/bbmap/bbduk/main.nf [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+workflow test_bbmap_bbduk_pe_ref {
+
+    input =  [  [ id:'test', single_end:false ], // meta map
+                [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
+                  file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
+             ]
+    contaminants = [file(params.test_data['sarscov2']['genome']['transcriptome_fasta'], checkIfExists: true) ]
+    use_contaminants = true
+
+    BBMAP_BBDUK ( input, contaminants, use_contaminants )
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:52]:
```suggestion
    BBMAP_BBDUK ( input, contaminants )
```

File: tests/software/bbmap/bbduk/main.nf [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+
+workflow test_bbmap_bbduk_pe_ref {
+
+    input =  [  [ id:'test', single_end:false ], // meta map
+                [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
+                  file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
+             ]
+    contaminants = [file(params.test_data['sarscov2']['genome']['transcriptome_fasta'], checkIfExists: true) ]
+    use_contaminants = true
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:50]:
```suggestion
```

File: tests/software/bbmap/bbduk/main.nf [hunk 16]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { BBMAP_BBDUK } from '../../../../software/bbmap/bbduk/main.nf' addParams( options: [ 'args' : 'trimq=10 qtrim=r' ] )
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:5]:
```suggestion
include { BBMAP_BBDUK } from '../../../../software/bbmap/bbduk/main.nf' addParams( options: [ 'args' : 'trimq=10 qtrim=r', 'suffix' : 'trim' ] )
```

**Review by drpatelh — CHANGES_REQUESTED** [review]
Looks great for a first stab @MGordon09! Thanks!

Some minor changes mainly to amend the module to take optional inputs by providing `[]` if the file isn't required. This isn't official yet but I believe it should work without needing to have an additional boolean parameter.

Ping me when you are happy with everything :)

File: tests/software/bbmap/bbduk/main.nf [hunk 17]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { BBMAP_BBDUK } from '../../../../software/bbmap/bbduk/main.nf' addParams( options: [ 'args' : 'trimq=10 qtrim=r', 'suffix' : 'trim' ] )
```
Comment (drpatelh) [inline tests/software/bbmap/bbduk/main.nf:5]:
```suggestion
include { BBMAP_BBDUK } from '../../../../software/bbmap/bbduk/main.nf' addParams( options: [ 'args' : 'trimq=10 qtrim=r', 'suffix' : '.trim' ] )
```

Comment (MGordon09) [reply tests/software/bbmap/bbduk/main.nf]: Hi @drpatelh,
Thank you so much for the guidance & feedback; I've added the suggested edits so hopefully it should be up to scratch now. Apologies for any silly errors - I'm a bit of an NF noob so very much learning as I go!
                                                                                                                                                                                                      
Thank you and the rest of the NF-core team for all your hard work developing and maintaining this great resource and if there is anything else you need me to change please give m a shout!

File: software/bbmap/bbduk/main.nf [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    input:
+    tuple val(meta), path(reads)
+    path contaminants
+
+    output:
+    tuple val(meta), path('*.fastq.gz'), emit: reads
+    tuple val(meta), path('*.log')          , emit: log
+    path '*.version.txt'                    , emit: version
```
Comment (drpatelh) [inline software/bbmap/bbduk/main.nf:27]:
```suggestion
    tuple val(meta), path('*.fastq.gz'), emit: reads
    tuple val(meta), path('*.log')     , emit: log
    path '*.version.txt'               , emit: version
```

**Review by drpatelh — APPROVED** [review]
Looks awesome @MGordon09 ! Great work for a first attempt :)

**Pull request merged**